templates/tftmpl/template_services.go

```diff
@@ -0,0 +1,161 @@
+package tftmpl
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/consul-terraform-sync/logging"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+)
+
+var (
+	_ Template = (*ServicesTemplate)(nil)
+)
+
+// ServicesTemplate handles the template for the services variable for the
+// template function: `{{ service }}`
+type ServicesTemplate struct {
+	Names []string
+
+	SourceIncludesVar bool
+
+	// Introduced in 0.5 - optional overall service filtering configured through
+	// the task's condition "services". These configs or Services can be
+	// configured but not both.
+	Datacenter string
+	Namespace  string
+	Filter     string
+
+	// Deprecated in 0.5 - optional per service filtering configured through the
+	// task's services list. Not all services configured in Names must have
+	// filtering configured. Services or the set of {Datacenter,
+	// Namespace, Filter} can be configured but not both.
+	Services map[string]Service
+}
+
+// Service contains additional Consul service filtering information for services
+// configured in ServicesTemplate
+type Service struct {
+	Datacenter string
+	Namespace  string
+	Filter     string
+}
+
+// IsServicesVar returns true because the template is for the services variable
+func (t ServicesTemplate) IsServicesVar() bool {
+	return true
+}
+
+func (t ServicesTemplate) appendModuleAttribute(*hclwrite.Body) {}
+
+func (t ServicesTemplate) appendTemplate(w io.Writer) error {
+	tmpl, err := t.concatServiceTemplates()
+	if err != nil {
+		return err
+	}
+
+	// note: ServicesTemplate currently only supports source_includes_var = true
+	// Wrap all templates in variable declaration `services = {}`
+	tmpl = fmt.Sprintf(servicesIncludesVarTmpl, tmpl)
+
+	if _, err := fmt.Fprint(w, tmpl); err != nil {
+		logging.Global().Named(logSystemName).Named(tftmplSubsystemName).Error(
+			"unable to write services template", "error", err)
+		return err
+	}
+	return nil
+}
+
+// concatServiceTemplate sorts the services alphabetically and concatenates
+// a template for each service
+func (t ServicesTemplate) concatServiceTemplates() (string, error) {
+	// double-check that service query parameter is configured in only one way
+	// the current way or the deprecated way
+	isCurrent := t.Datacenter != "" || t.Namespace != "" || t.Filter != ""
+	isDeprecated := t.Services != nil
+
+	if isCurrent && isDeprecated {
```

Comment:
Appreciate the readability here! Just double checking, this should never occur with validation done earlier in the code path?

Reply:
Yes, this is not possible for the user due to config validation - added a comment to hopefully clarify. It is possible if ServicesTemplate is manually configured with both e.g. in a  test-case

Reply:
Thanks for the info! Oh sweet I see a new comment added for context too.